javalib/src/main/scala/java/math/BigInteger.scala

```diff
@@ -0,0 +1,1500 @@
+/*
+ *  Ported by Alistair Johnson from  https://github.com/gwtproject/gwt/blob/master/user/super/com/google/gwt/emul/java/math/BigInteger.java
+ */
+package java.math
+
+import java.io.Serializable
+import java.util.Random
+
+import scala.annotation.tailrec
+
+/**
+ * Companion object for {@code BigInteger}
+ */
+object BigInteger {
+
+  /**
+   * The {@code BigInteger} constant 1.
+   */
+  val ONE = new BigInteger(1, 1)
+
+  /**
+   * The {@code BigInteger} constant 10.
+   */
+  val TEN = new BigInteger(1, 10)
+
+  /**
+   * The {@code BigInteger} constant 0.
+   */
+  val ZERO = new BigInteger(0, 0)
+
+  /**
+   * The {@code BigInteger} constant 0 used for comparison.
+   */
+  val EQUALS = 0
+
+  /**
+   * The {@code BigInteger} constant 1 used for comparison.
+   */
+  val GREATER = 1
+
+  /**
+   * The {@code BigInteger} constant -1 used for comparison.
+   */
+  val LESS = -1
+
+  /**
+   * The {@code BigInteger} constant -1.
+   */
+  val MINUS_ONE = new BigInteger(-1, 1)
+
+  /**
+   * 2^32.
+   */
+  val POW32 = 4294967296d
+
+  /**
+   * All the {@code BigInteger} numbers in the range [0,10] are cached.
+   */
+  val SMALL_VALUES = Array(
+    ZERO, ONE, new BigInteger(1, 2), new BigInteger(1, 3),
+    new BigInteger(1, 4), new BigInteger(1, 5), new BigInteger(1, 6),
+    new BigInteger(1, 7), new BigInteger(1, 8), new BigInteger(1, 9), TEN)
+
+  val TWO_POWS = new Array[BigInteger](32)
+
+  for (i <- 0 until TWO_POWS.length) {
+    TWO_POWS(i) = BigInteger.valueOf(1L << i)
+  }
+
+  /**
+   * Returns a random positive {@code BigInteger} instance in the range [0,
+   * 2^(bitLength)-1] which is probably prime. The probability that the returned
+   * {@code BigInteger} is prime is beyond (1-1/2^80).
+   * <p>
+   * <b>Implementation Note:</b> Currently {@code rnd} is ignored.
+   *
+   * @param bitLength length of the new {@code BigInteger} in bits.
+   * @param rnd random generator used to generate the new {@code BigInteger}.
+   * @return probably prime random {@code BigInteger} instance.
+   * @throws ArithmeticException if {@code bitLength < 2}.
+   */
+  def probablePrime(bitLength: Int, rnd: Random): BigInteger = new BigInteger(bitLength, 100, rnd)
+
+  def valueOf(lVal: Long): BigInteger = {
+    if (lVal < 0) {
+      if (lVal != -1) {
+        new BigInteger(-1, -lVal)
+      } else {
+        MINUS_ONE
+      }
+    } else if (lVal <= 10) {
+      SMALL_VALUES(lVal.toInt)
+    } else {
+      new BigInteger(1, lVal)
+    }
+  }
+
+  def getPowerOfTwo(exp: Int): BigInteger = {
```

Comment:
private

In general, private-ize all methods that are not part of the public API in the JDK. If necessary, make them `private[math]`, but they shouldn't be `public`.